lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexMaster.java

```diff
@@ -107,37 +111,78 @@ public void workerPreempted(final String id) {
-   */
   @Override
-  public void taskletCompleted(final String workerId,
-                               final int taskletId,
-                               final Serializable result) {
-    runningWorkers.completeTasklet(workerId, taskletId, result);
+  public void workerReported(final String workerId, final WorkerReport workerReport) {
+    // TODO[JIRA REEF-942]: Fix when aggregation is allowed.
+
+    for (final TaskletReport taskletReport : workerReport.getTaskletReports()) {
+      switch (taskletReport.getType()) {
+      case TaskletResult:
+        final TaskletResultReport taskletResultReport = (TaskletResultReport) taskletReport;
+
+        final List<Integer> resultTaskletIds = taskletResultReport.getTaskletIds();
+        runningWorkers.doneTasklets(workerId, resultTaskletIds);
+        fetchDelegate(resultTaskletIds).completed(resultTaskletIds, taskletResultReport.getResult());
+
+        break;
+      case TaskletCancelled:
+        final TaskletCancelledReport taskletCancelledReport = (TaskletCancelledReport) taskletReport;
+        final List<Integer> cancelledIdToList = Collections.singletonList(taskletCancelledReport.getTaskletId());
+        runningWorkers.doneTasklets(workerId, cancelledIdToList);
+        fetchDelegate(cancelledIdToList).cancelled(taskletCancelledReport.getTaskletId());
+
+        break;
+      case TaskletFailure:
+        final TaskletFailureReport taskletFailureReport = (TaskletFailureReport) taskletReport;
+
+        final List<Integer> failureTaskletIds = taskletFailureReport.getTaskletIds();
+        runningWorkers.doneTasklets(workerId, taskletFailureReport.getTaskletIds());
+        fetchDelegate(failureTaskletIds).threwException(failureTaskletIds, taskletFailureReport.getException());
+
+        break;
+      default:
+        throw new RuntimeException("Unknown Report");
+      }
+    }
   }
 
   /**
-   * Notify task failure to runningWorkers.
+   * Terminate the job.
    */
   @Override
-  public void taskletErrored(final String workerId, final int taskletId, final Exception exception) {
-    runningWorkers.errorTasklet(workerId, taskletId, exception);
+  public void terminate() {
+    runningWorkers.terminate();
   }
 
   /**
-   * Notify tasklet cancellation to runningWorkers.
+   * Puts a delegate to associate with a Tasklet.
    */
-  @Override
-  public void taskletCancelled(final String workerId, final int taskletId) {
-    runningWorkers.taskletCancelled(workerId, taskletId);
+  private synchronized void putDelegate(final List<Tasklet> tasklets, final VortexFutureDelegate delegate) {
+    for (final Tasklet tasklet : tasklets) {
+      taskletFutureMap.put(tasklet.getId(), delegate);
+    }
   }
 
   /**
-   * Terminate the job.
+   * Fetches a delegate that maps to the list of Tasklets.
    */
-  @Override
-  public void terminate() {
-    runningWorkers.terminate();
+  private synchronized VortexFutureDelegate fetchDelegate(final List<Integer> taskletIds) {
```

Comment:
I was thinking of a new class `TaskletGroup` that includes a list of `Tasklet`s that's locally aggregated at a worker. A `TaskletGroup` is mapped to a `VortexFutureDelegate`.